libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -835,67 +842,527 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+    for (uint8_t i=0; i<ARRAY_SIZE(deferred_message_bucket); i++) {
+        const deferred_message_bucket_t &bucket = deferred_message_bucket[i];
+        if (bucket.interval_ms == 0) {
+            // unused bucket
+            if (empty_bucket_id == -1) {
+                empty_bucket_id = i;
+            }
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            if (bucket.ap_message_ids.first_set() != -1) {
+                ::fprintf(stderr, "Bucket %u has zero interval but with ids set", i);
+                abort();
+            }
+#endif
+            continue;
+        }
+        if (bucket.ap_message_ids.get(id)) {
+            in_bucket = i;
+        }
+        const uint16_t interval_delta = abs(bucket.interval_ms - interval_ms);
+        if (interval_delta < closest_bucket_interval_delta) {
+            closest_bucket = i;
+            closest_bucket_interval_delta = interval_delta;
         }
     }
 
-    // we failed to send the message this time around, so try to defer:
-    if (num_deferred_messages == ARRAY_SIZE(deferred_messages)) {
-        // the defer buffer is full, discard this attempt to send.
-        // Note that the message *may* already be in the defer buffer
-        return;
+    if (in_bucket == -1 && interval_ms == 0) {
+        // not in a bucket and told to remove from scheduling
+        return true;
     }
 
-    // check if this message is deferred:
-    for (i=0, nextid = next_deferred_message; i < num_deferred_messages; i++) {
-        if (deferred_messages[nextid] == id) {
-            // it's already deferred
-            return;
+    if (in_bucket != -1) {
+        if (interval_ms == 0) {
+            // remove it
+            remove_message_from_bucket(in_bucket, id);
+            return true;
         }
-        nextid++;
-        if (nextid == ARRAY_SIZE(deferred_messages)) {
-            nextid = 0;
+        if (closest_bucket_interval_delta == 0 &&
+            in_bucket == closest_bucket) {
+            // don't need to move it
+            return true;
         }
+        // remove from existing bucket
+        remove_message_from_bucket(in_bucket, id);
+        if (empty_bucket_id == -1 &&
+            deferred_message_bucket[in_bucket].ap_message_ids.count() == 0) {
+            empty_bucket_id = in_bucket;
+        }
+    }
+
+    if (closest_bucket == -1 && empty_bucket_id == -1) {
+        // gah?!
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        ::fprintf(stderr, "no buckets?!\n");
+        abort();
+#endif
+        return false;
+    }
+
+    if (closest_bucket_interval_delta > 5 &&
+        empty_bucket_id != -1) {
+        // allocate a bucket for this interval
+        deferred_message_bucket[empty_bucket_id].interval_ms = interval_ms;
+        deferred_message_bucket[empty_bucket_id].last_sent_ms = AP_HAL::millis16();
+        closest_bucket = empty_bucket_id;
+        closest_bucket_interval_delta = 0;
+    } else if (deferred_message_bucket[closest_bucket].ap_message_ids.count() == 0) {
```

Comment:
@peterbarker Not saying the contrary, I just can't understand when this would be hit. If a bucket has zero message ids then it has zero interval then it is an empty bucket and then it isn't a closest bucket.